### PR TITLE
rPackages: mark packages as broken when generating package sets

### PR DESCRIPTION
**Pull request by jbedo**
The current behaviour for generate-r-packages.R is to delete
packages that have been remove upstream. This patch changes the
behaviour to mark packages as broken rather than removing them.
This has the advantage of never breaking expressions, which
previously occured when a package with overrides in default.nix
was deleted. As a result, the update procedure is simplified,
allowing automated updates to the package tree to run, and
additionally if a package is re-established upstream the previous
overrides still exist.

@PhDyellow we were discussing something similiar.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by teto** [comment]
On an unrelated topic, I've changed the luarocks update script (similar to this one) to leverage the code used by pkgs/misc/vim-plugins/update.py . I wonder if you would be interested to do (or let anyone do) the same ? The advantage is that updater scripts provide the same CLI interface across ecosystems, with `--autocommit` capabilities, the ability to choose the input/output filename so that everyone can maintain their own overlay etc

**Comment by jbedo** [comment]
I am interested in unifying across ecosystems, but there are some
particulars with the R package repositories and I wonder how to handle
them with the vim-plugins update code. Firstly, the package index does
not provide sha256 hashes only md5, so we currently fetch each package
and hash it ourselves. Secondly the BIOC releases are against specific R
versions, and AFAIK the map between them is only queriable via the
BiocManager R package.

File: doc/languages-frameworks/r.section.md [hunk 1]
```diff
@@ -113,11 +113,3 @@ mv bioc-experiment-packages.nix.new bioc-experiment-packages.nix
...
 
 `generate-r-packages.R <repo>` reads  `<repo>-packages.nix`, therefor the renaming.
-
-## Testing if the Nix-expression could be evaluated {#testing-if-the-nix-expression-could-be-evaluated}
-
-```bash
-nix-build test-evaluation.nix --dry-run
-```
-
-If this exits fine, the expression is ok. If not, you have to edit `default.nix`
```
Comment (teto) [inline doc/languages-frameworks/r.section.md:123]: I think it's a good idea to mention in which files we put overrides, but with the fullpath.

**Comment by teto** [comment]
maintainers/scripts/pluginupdate.py defines an abstract Editor class (I had hoped to rename it to Ecosystem) that can be derived for specific ecosystems, it is currently used in vim/kakoune/lua (maintainers/scripts/update-luarocks-packages) and could be implemented for R. Little is assumed so  it should still be possible to implement what you mentioned (converting from md5 to sha256). For instance the lua generator adds to each package the lua versions they are valid for. 

I dont use R so I would rather let someone using it merge. Looks ok.

**Comment by jbedo** [comment]
Great, I'll take a look at it. My priority atm is to get the current
tree in a more robust state where we have regular automatic updates and
hydra builds. We're not too far off this and I can take a look at
refactoring using other frameworks afterwards.

Good suggestion to mention where overrides go, I'll add a section to the
docs.

File: doc/languages-frameworks/r.section.md (hunk 1 above)
Comment (jbedo) [reply doc/languages-frameworks/r.section.md:123]: Added text describing overrides to docs in latest commit.

**Comment by jbedo** [comment]
This has been working for days now on my automatically updated R branch. Pull request at #139609 updating the package set.

I think we can merge this if everybody is happy with the docs.

**Comment by jbedo** [comment]
#140747 just got merged and we have hydra building the rPackages tree. We can now work on the nixpkgs/r-updates and fix all the errors :).

**Comment by jbedo** [comment]
BTW for the failing home check on `styler` there's #129746

**Pull request merged**